csrc/multidevice/utils.cpp

```diff
@@ -345,22 +371,49 @@ bool haveDifferentShardings(
   // Therefore, we collect all the loop IterDomains that depend on the
   // logical-domain-mapped IterDomains, and check if they are DID-parallelized
   // consistently.
-  const std::unordered_map<IterDomain*, IterDomain*>& p2c =
-      PairwiseLogicalDomainMap(producer, consumer).mapProducerToConsumer();
-  std::unordered_set<IterDomain*> mapped_p_logical_ids;
-  mapped_p_logical_ids.reserve(p2c.size());
-  std::unordered_set<IterDomain*> mapped_c_root_ids;
-  mapped_c_root_ids.reserve(p2c.size());
-  for (IterDomain* p_logical_id : producer->getLogicalDomain()) {
-    const auto i = p2c.find(p_logical_id);
-    if (i == p2c.end()) {
-      // This happens e.g. when `p_logical_id` is squeezed or is a product of a
-      // reduction. Even if `p_logical_id` is parallelized on DID, the
-      // dimension is size-1 and doesn't trigger resharding.
+  const std::unordered_map<IterDomain*, IterDomain*>& c2p =
```

Comment:
Do we need to worry about broadcast IDs here? For historical reasons, broadcast and corresponding concrete IDs are mapped by default in `PairwiseLogicalDomainMap`. To disable that, add `mapBroadcast(false)` before calling `mapConsumerToProducer`.

Reply:
Good point. ReshardingTest.Add_Broadcast does pass so I haven't thought too much about broadcasts. I'll try mapBroadcast(false) to see if it triggers new errors.

Reply:
Done

Reply:
If it's b vs non-b, mapBroadcast skips them so they don't affect the result of isResharding. IIUC, this is the correct behavior. 

If it's b vs b, the existing logic kicks in to check how they are sharded. Therefore, `b{1}` vs `bDIDx{1}` will be treated as resharding. This is a bit problematic because we could consider them non-resharding and save communication. I [tried](https://github.com/NVIDIA/Fuser/pull/3053) that idea a long time ago but it didn't even get past the pointwise scheduler. IIRC, the symptom is that the pointwise scheduler tried to split `bDIDx{1}` in a reference tensor and ended up with multiple `DIDx`s in the loop domain, which is invalid. However, I'm happy to revisit that after this PR.

Reply:
I was thinking about the first case. I'm not sure what you mean they don't affect the result. IIUC, `c2p` is only used at line 421, so for a pair of a broadcast producer ID and a non-broadcast consumer ID, two different index vals are created. Wouldn't they cause mismatched `p_index` and `c_index`?

Reply:
> two different index vals are created

That's right. 

> c2p is only used at line 421

Not exactly. c2p is also indirectly used here https://github.com/NVIDIA/Fuser/pull/3482/files#diff-4739b2bc50e3c078541f1ea6665370ea6dd853d1fa0b4d1085b95d7c4791d96dR457 and later for c_id as well. This makes `p_index == c_index == nullptr`. This is intentional because I do want to skip non-mapped IDs (cf. https://github.com/NVIDIA/Fuser/blob/5da602dc3dc08b8a05d818eb9376670d8c3227e4/csrc/multidevice/utils.cpp#L329-L331)

Reply:
Ah, OK, that makes sense.